### ISSUE TITLE
[JIT] Question: typeof(T).Name doesn't become a constant?

### ISSUE THREAD

**Issue by CodeBlanch** [dotnet/runtime#81564]
I was working on some code where I wanted to push `typeof(T).Name` in a generic method as an attribute on some log message. I just assumed it would become a cheapish constant thing but apparently does not:

![image](https://user-images.githubusercontent.com/28367120/216448707-8c8fffc3-393a-4072-9748-74145b51c4c6.png)

Just curious if there is a reason for that? Perhaps an optimization opportunity?

**Comment by EgorBo** [comment]
It shouldn't be difficult to add this in JIT, but as far as I remember `type(T).Name` is not that expensive - it will load the string from some cache/dictionary. Do you have some real world example where an inlined string literal would improve things/fold something?

**Comment by CodeBlanch** [comment]
Here's a benchmark that is close to what I'm doing:

```csharp
public class LogBenchmarks
{
    private readonly SomeClass<int> instance = new();

    [Benchmark]
    public void LogTypeof()
    {
        this.instance.SomeLogicTypeofVersion();
    }

    [Benchmark]
    public void LogUsingField()
    {
        this.instance.SomeLogicFieldVersion();
    }

    private sealed class SomeClass<T>
    {
        private readonly string typeName;

        public SomeClass()
        {
            this.typeName = typeof(T).Name;
        }

        public void SomeLogicTypeofVersion()
        {
            LogBenchmarksEventSource.Log.WriteMessage(typeof(T).Name);
        }

        public void SomeLogicFieldVersion()
        {
            LogBenchmarksEventSource.Log.WriteMessage(this.typeName);
        }
    }

    [EventSource(Name = "LogBenchmarks")]
    private sealed class LogBenchmarksEventSource : EventSource
    {
        public static LogBenchmarksEventSource Log { get; } = new LogBenchmarksEventSource();

        [Event(1, Message = "Some log message for type '{0}'.", Level = EventLevel.Informational)]
        public void WriteMessage(string itemType)
        {
            this.WriteEvent(1, itemType);
        }
    }
}
```

The EventSource writes will no-op unless someone is listening so I'm just trying to keep them as low overhead as possible.

|        Method |      Mean |     Error |    StdDev |
|-------------- |----------:|----------:|----------:|
|     LogTypeof | 3.9900 ns | 0.0549 ns | 0.0513 ns |
| LogUsingField | 0.9410 ns | 0.0429 ns | 0.0401 ns |

**Comment by EgorBo** [comment]
This will likely need a new JIT-VM API (quote verbose) and the pattern itself doesn't seem popular https://grep.app/search?q=typeof%5C%28.%2A%5C%29.Name&regexp=true&filter[lang][0]=C%23 - only 22 hits

**Comment by dn9090** [comment]
> This will likely need a new JIT-VM API (quote verbose) and the pattern itself doesn't seem popular [https://grep.app/search?q=typeof%5C%28.%2A%5C%29.Name&regexp=true&filter[lang][0]=C%23](https://grep.app/search?q=typeof%5C%28.%2A%5C%29.Name&regexp=true&filter%5Blang%5D%5B0%5D=C%23) - only 22 hits
> 
> Also, I am not sure what VM offers for type name matches C# Type.Name for generics

I just want to note that I also use the type name to generate a stable hash code (e.g. `var hashCode = GetHashCode(type.Name)`. For me, this optimization would also be beneficial.

**Comment by EgorBo** [comment]
> For me, this optimization would also be beneficial.

GetHashCode is not foldable to const strings (mainly because it relies on per-process random salt) so in your case you still will have to spend a lot of cycles inside it so `type.Name` unlikely to make a difference IMO

**Comment by dn9090** [comment]
> > For me, this optimization would also be beneficial.
> 
> GetHashCode is not foldable to const strings (mainly because it relies on per-process random salt) so in your case you still will have to spend a lot of cycles inside it so `type.Name` unlikely to make a difference IMO

This is true when I need to regenerate the hash. But for me it is important that the length of the string is also a constant. So the compiler should drop out unnecessary branches (in the hash function), which needed a certain length of the string right? Like:
```
var name = typeof(Klass).Name;
if(name.Length < 8) FastPath();
else SlowPath();
```

**Comment by EgorBo** [comment]
> 

is your hashcode faster than the default Type.GetHashCode ?

**Comment by MichalPetryka** [comment]
I feel like usually `typeof(T).FullName` is the more comon version.

**Comment by dn9090** [comment]
> is your hashcode faster than the default Type.GetHashCode ?

I do not know. Never came into question for me because the hash code is not stable.
